File: uc-08-evaluacion-testing/src/evaluation/golden_dataset.py
"""
Loader y validador del golden dataset compartido con RAGAS y DeepEval.

Equivalente Python de goldenDataset.ts — leen el mismo data/seeds/golden-dataset.json
para que toda la suite (TS + Python) trabaje sobre la misma fuente de verdad.

Uso:
    from golden_dataset import load_golden, filter_rag_pairs
    dataset = load_golden()
    pairs = filter_rag_pairs(dataset)
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

# Resolver respecto a la raíz del repo (4 niveles arriba):
# uc-08-evaluacion-testing/src/evaluation/golden_dataset.py
REPO_ROOT = Path(__file__).resolve().parents[3]
GOLDEN_PATH = REPO_ROOT / "data" / "seeds" / "golden-dataset.json"
CATALOGO_PATH = REPO_ROOT / "data" / "seeds" / "propiedades.json"

# Categorías donde expectedPropertyIds = [] es válido (preguntas conceptuales)
CATEGORIES_ALLOW_EMPTY = {"fuera-de-rango", "conversacional"}


def load_golden() -> list[dict[str, Any]]:
    """Carga las 20 preguntas. Lanza FileNotFoundError si el seed no está generado."""
    with GOLDEN_PATH.open(encoding="utf-8") as f:
        data = json.load(f)
    return data["preguntas"]


def load_catalog_ids() -> set[str]:
    """IDs de propiedades del catálogo — para validar referencias del dataset."""
    with CATALOGO_PATH.open(encoding="utf-8") as f:
        return {p["id"] for p in json.load(f)}


def filter_rag_pairs(dataset: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """
    Devuelve solo las preguntas que sirven para evaluar RAG:
    - tienen propiedades esperadas (expectedPropertyIds no vacío)
    - no son conceptuales (fuera-de-rango, conversacional)
    """
    return [p for p in dataset if p["expectedPropertyIds"] and p["categoria"] not in CATEGORIES_ALLOW_EMPTY]


def filter_conceptual(dataset: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Preguntas conceptuales — se evalúan con métricas de comportamiento defensivo."""
    return [p for p in dataset if p["categoria"] in CATEGORIES_ALLOW_EMPTY]


if __name__ == "__main__":
    dataset = load_golden()
    rag_pairs = filter_rag_pairs(dataset)
    conceptual = filter_conceptual(dataset)
    print(f"Total preguntas: {len(dataset)}")
    print(f"  Para RAG (RAGAS):       {len(rag_pairs)}")
    print(f"  Conceptuales/fuera:     {len(conceptual)}")
    print("\nMuestras:")
    for q in dataset[:3]:
        print(f"  [{q['categoria']}] {q['question']}")

File: uc-08-evaluacion-testing/src/evaluation/goldenDataset.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { z } from 'zod';

/**
 * Schema del golden dataset que evalúan RAGAS, DeepEval y Promptfoo.
 * Validar el dataset en TS antes de correr cualquier evaluación con LLM
 * evita gastar tokens descubriendo que el dataset estaba mal formado.
 */
export const PreguntaGoldenSchema = z.object({
  id: z.string().regex(/^q-\d{3,}$/),
  categoria: z.enum([
    'busqueda-semantica',
    'filtros-mixtos',
    'preferencias-comprador',
    'preferencias-familia',
    'negocio-comercial',
    'negocio-vendedor',
    'presupuesto',
    'ubicacion-especifica',
    'amenidades',
    'fuera-de-rango',
    'conversacional',
  ]),
  question: z.string().min(10),
  expectedPropertyIds: z.array(z.string()),
  idealAnswer: z.string().min(20),
  contexts: z.array(z.string()),
});

export const GoldenDatasetSchema = z.object({
  version: z.string(),
  descripcion: z.string(),
  preguntas: z.array(PreguntaGoldenSchema).min(1),
});

export type PreguntaGolden = z.infer<typeof PreguntaGoldenSchema>;
export type GoldenDataset = z.infer<typeof GoldenDatasetSchema>;

const GOLDEN_PATH = resolve(process.cwd(), 'data/seeds/golden-dataset.json');
const CATALOGO_PATH = resolve(process.cwd(), 'data/seeds/propiedades.json');

export function loadGoldenDataset(): GoldenDataset {
  const raw = JSON.parse(readFileSync(GOLDEN_PATH, 'utf-8'));
  return GoldenDatasetSchema.parse(raw);
}

export function loadCatalogIds(): Set<string> {
  const raw = JSON.parse(readFileSync(CATALOGO_PATH, 'utf-8')) as Array<{ id: string }>;
  return new Set(raw.map((p) => p.id));
}

/**
 * Conjunto de barrios/zonas presentes en el catálogo. Útil para validar que
 * los `idealAnswer` mencionan alguna zona reconocible (anchor de faithfulness).
 * Se extrae del catálogo dinámicamente en vez de hardcodear — así crece con los seeds.
 */
export function loadCatalogBarrios(): Set<string> {
  const raw = JSON.parse(readFileSync(CATALOGO_PATH, 'utf-8')) as Array<{
    ubicacion?: { barrio?: string; zona?: string; ciudad?: string };
  }>;
  const set = new Set<string>();
  for (const p of raw) {
    if (p.ubicacion?.barrio) set.add(p.ubicacion.barrio);
    if (p.ubicacion?.zona) set.add(p.ubicacion.zona);
  }
  return set;
}
